### ISSUE TITLE
3-bit quantization relative error ~18% (expected ~5%), causes garbled output on Gemma4

### ISSUE THREAD

**Issue by labolado** [arozanov/turboquant-mlx#3]
## Environment
- turboquant-mlx 0.2.0
- mlx 0.31.1, mlx-lm 0.31.1, mlx-vlm 0.4.4
- Apple Silicon (M5 Pro)
- Model: gemma-4-26b-a4b-it-4bit

## Problem

Quantization relative error is much higher than expected from the TurboQuant paper (arXiv 2504.19874).

**Test:**
```python
from turboquant_mlx import TurboQuantKVCache
import mlx.core as mx

B, H, S, D = 1, 8, 10, 256
keys = mx.random.normal((B, H, S, D)).astype(mx.float16)
values = mx.random.normal((B, H, S, D)).astype(mx.float16)

for bits in [2, 3, 4]:
    cache = TurboQuantKVCache(bits=bits)
    k_out, v_out = cache.update_and_fetch(keys, values)
    k_rel = (mx.mean(mx.abs(keys - k_out)) / mx.mean(mx.abs(keys))).item()
    print(f'{bits}-bit: K_rel_err={k_rel:.4f}')
```

**Results:**
```
2-bit: K_rel_err=0.3408
3-bit: K_rel_err=0.1864
4-bit: K_rel_err=0.0977
```

The paper reports ~1-2% error for 3-bit turbo3 scheme. The 18.6% relative error here causes garbled output when used as `prompt_cache` in mlx-vlm:

```
Output: Here-:-로-0-1-1-&-&$-0-1-&--111-1111111-111-111-0اً-0-0{-111111111로...
```

Error is consistent across dimensions (D=64/128/256) and independent of the model.

## Additional fix needed

`TurboQuantKVCache.make_mask()` references undefined `create_attention_mask`. Fix:
```python
def make_mask(self, N, return_array=False, window_size=None):
    if N == 1:
        return None
    from mlx_lm.models.base import create_causal_mask
    if return_array or (window_size and N > window_size):
        return create_causal_mask(N, offset=self.offset, window_size=window_size)
    return "causal"
```

**Comment by labolado** [comment]
Update: I found a bug in my testing methodology — the baseline was also producing garbled output because I was reusing the model instance across multiple runs without clearing internal state. After fixing:

- **Baseline (no TQ)**: Perfect output ✅
- **TQ-all-30layers-3bit**: Semi-coherent but garbled — you can see topic keywords ("hash-table", "keys-with-values") but token selection is off. Speed doubles (30→60 tok/s).
- **TQ-global-only (5 layers)**: Fully garbled ❌

The element-wise error metric I reported (18%) was misleading — attention score MAE is 0.004 which looks reasonable. But the end-to-end generation quality is still degraded at 3-bit on this model.

Could this be a quantizer calibration issue? The Lloyd-Max boundaries may not be optimal for Gemma4's KV distribution.

**Comment by labolado** [comment]
Further investigation — the 18% element-wise error is confirmed to degrade generation quality. Clean single-process test:

```
Baseline: A hash table is a data structure that organizes data into key-value pairs...  ✅
TQ-3bit:  A-hash-table으로-an-organized-collection-of-keys-and-values-designed...  ❌
TQ-4bit:  A-hash table으로 than-ing-en else-way-of-describing-it-is-a-data-structure...  ⚠️
```

4-bit is noticeably better than 3-bit but still has artifacts.

**Root cause hypothesis**: The quantize kernel does NOT apply `1/sqrt(dim)` after the WHT butterfly, so the WHT output has std ≈ sqrt(256) = 16, while the Gaussian Lloyd-Max codebook only covers [-2.15, 2.15]. Most values get clipped to boundary centroids, losing precision.

The dequant kernel applies `scale = 1/sqrt(dim)` twice (once on centroid lookup, once after inverse WHT), which compensates for reconstruction magnitude but doesn't recover the clipped information.

I tried adding `1/sqrt(dim)` to the quantize kernel (to bring WHT output into codebook range) and removing the first `scale` from dequant (to keep only one `1/sqrt(dim)` for inverse WHT normalization), but this made things worse — suggesting the current double-scale in dequant is intentional compensation for the un-normalized WHT.

The fix might need to be in the codebook itself — scale the Lloyd-Max centroids by `sqrt(dim)` to match the un-normalized WHT output range, then adjust dequant accordingly.
